public/app/plugins/datasource/grafana-azure-monitor-datasource/composable_dataquery.cue

```diff
@@ -0,0 +1,188 @@
+// Copyright 2023 Grafana Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grafanaplugin
+
+import (
+	common "github.com/grafana/grafana/packages/grafana-schema/src/common"
+	"github.com/grafana/grafana/pkg/plugins/pfs"
+)
+
+// This file (with its sibling .cue files) implements pfs.GrafanaPlugin
+pfs.GrafanaPlugin
+
+composableKinds: DataQuery: {
+	maturity: "merged"
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						#AzureMonitorQuery: common.DataQuery & {
+							subscription?: string
+
+							// ARG uses multiple subscriptions
+							subscriptions?: [...string]
+
+							azureMonitor?:              #AzureMetricQuery
+							azureLogAnalytics?:         #AzureLogsQuery
+							azureResourceGraph?:        #AzureResourceGraphQuery
+							grafanaTemplateVariableFn?: #GrafanaTemplateVariableQuery
+
+							// Template variables params 
+							resourceGroup?: string
+							namespace?:     string
+							resource?:      string
+							query?:         #AzureQueryType
+						} @cuetsy(kind="interface")
+
+						// GrafanaTemplateVariableFn is deprecated
+						#AzureQueryType: "Azure Monitor" | "Azure Log Analytics" | "Azure Resource Graph" | "Azure Subscriptions" | "Azure Resource Groups" | "Azure Namespaces" | "Azure Resource Names" | "Azure Metric Names" | "Azure Workspaces" | "Grafana Template Variable Function" @cuetsy(kind="enum", memberNames="AzureMonitor|LogAnalytics|AzureResourceGraph|SubscriptionsQuery|ResourceGroupsQuery|NamespacesQuery|ResourceNamesQuery|MetricNamesQuery|WorkspacesQuery|GrafanaTemplateVariableFn")
+
+						// Azure Monitor Metrics sub-query properties
+						#AzureMetricQuery: {
+							resources?: [...#AzureMonitorResource]
```

Comment:
We should probably define these as mandatory. I _think_ that these are optional in the frontend because it's waiting for user input to fill certain fields. This is not really necessary AFAIK. A working query should have `resources`, `metricNamespace`, `metricName`... The model should probably reflect that (and leave as optional really the fields that can be empty).

Reply:
Agreed on this, I think it's time to take the opportunity and refactor the types in Azure Monitor as a part of the schematisation 😊

Reply:
> I think that these are optional in the frontend because it's waiting for user input to fill certain fields. 

This strikes right at the heart of one of the really interesting changes inherent in moving to a schema-centric model - when it comes to defaults and optionality, we have to think about object lifecycle quite differently. i wrote a bit about it [here](https://docs.google.com/document/d/1RRPGeD7DQ7wot0Pwnc4UZolXrZYip0ViqOozl9zbBpE/edit#heading=h.fft1p4jlnh9y), but it's going to be an evolving area.

And yes in this case i agree that "waiting for frontend user input" is not a sufficient reason to mark a field as optional